rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/RMController.java

```diff
@@ -506,172 +510,150 @@ public void onFailure(Throwable caught) {
-                long lockTime = getJsonLongNullable(nodeObj, "lockTime", -1);
-                String nodeLocker = getJsonStringNullable(nodeObj, "nodeLocker");
-
-                if (isLocked) {
-                    numLocked++;
-                }
+                final NodeSource nodeSource = newNodeSources.get(node.getSourceName());
 
-                Node n = new Node(nodeUrl,
-                                  nodeState,
-                                  nodeInfo,
-                                  timeStamp,
-                                  timeStampFormatted,
-                                  nodeProvider,
-                                  nodeOwner,
-                                  nss,
-                                  hostName,
-                                  vmName,
-                                  description,
-                                  defaultJMXUrl,
-                                  proactiveJMXUrl,
-                                  isLocked,
-                                  lockTime,
-                                  nodeLocker);
-
-                // deploying node
-                if (hostName == null || hostName.length() == 0) {
-                    ns.get(nss).getDeploying().put(nodeUrl, n);
-                }
-                // already deployed node
-                else {
+                // if node source was not deleted
+                if (nodeSource != null) {
 
-                    Host host = ns.get(nss).getHosts().get(hostName);
-                    if (host == null) {
-                        host = new Host(hostName, nss);
-                        ns.get(nss).getHosts().put(hostName, host);
-                    }
-                    host.getNodes().put(nodeUrl, n);
-                    if (nodeUrl.toLowerCase().contains("virt-")) {
-                        host.setVirtual(true);
+                    if (!node.isRemoved()) {
+                        addNodeToNodeSource(node, nodeSource);
+                    } else {
+                        removeNodeFromNodeSource(node, nodeSource);
                     }
                 }
 
-                switch (n.getNodeState()) {
-                    case BUSY:
-                        numBusy++;
-                        break;
-                    case CONFIGURING:
-                        numConfiguring++;
-                        break;
-                    case DEPLOYING:
-                        numDeploying++;
-                        break;
-                    case DOWN:
-                        numDown++;
-                        break;
-                    case FREE:
-                        numFree++;
-                        break;
-                    case LOST:
-                        numLost++;
-                        break;
-                    case TO_BE_REMOVED:
-                        numToBeRemoved++;
-                        break;
-                }
             } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)